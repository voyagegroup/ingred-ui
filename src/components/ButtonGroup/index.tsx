export { default, ButtonGroupProps } from "./ButtonGroup";
