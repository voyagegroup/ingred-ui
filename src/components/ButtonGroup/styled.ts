import styled from "styled-components";

export type ContainerProps = {
  minWidth: string;
  horizontalPadding: string;
};

export const ButtonGroupContainer = styled.div<ContainerProps>`
  display: inline-flex;

  & > * {
    min-width: ${({ minWidth }) => minWidth};
    width: auto;
    padding-right: ${({ horizontalPadding }) => horizontalPadding};
    padding-left: ${({ horizontalPadding }) => horizontalPadding};
  }

  & > *:not(:last-child) {
    border-bottom-right-radius: 0;
    border-top-right-radius: 0;
  }

  & > *:not(:first-child) {
    border-left: none;
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
  }

  & > *:last-child {
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
  }
`;
