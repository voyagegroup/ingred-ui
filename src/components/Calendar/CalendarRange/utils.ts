import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
