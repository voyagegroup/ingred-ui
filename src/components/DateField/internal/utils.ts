export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
