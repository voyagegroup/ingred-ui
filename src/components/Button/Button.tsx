import * as React from "react";
import { fontSize } from "../Typography/Typography";

import { Props as BaseButtonProps } from "./internal/BaseButton";
import * as Styled from "./styled";
import { Theme, useTheme } from "../../themes";
import { hexToRgba } from "../../utils/hexToRgba";

export type ButtonSize = "small" | "medium" | "large";
export type ButtonColor = "primary" | "secondary" | "danger" | "clear";

export type ButtonColorStyle = {
  normal: {
    background: string;
    color: string;
    boxShadow: string;
    border: string;
  };
  hover: {
    background?: string;
    border: string;
  };
  active: {
    background?: string;
    boxShadow: string;
    border: string;
  };
};

export type Padding = {
  theme: Theme;
  size: ButtonSize;
  color: ButtonColor;
};

const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: `0px -2px ${hexToRgba(
        theme.palette.black,
        0.16,
      )} inset, 0px 2px ${hexToRgba(theme.palette.black, 0.08)}`,
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: `inset 0 2px ${hexToRgba(theme.palette.black, 0.16)}`,
      border: "none",
    },
  },
  secondary: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: `0px -2px ${hexToRgba(
        theme.palette.black,
        0.16,
      )} inset, 0px 2px ${hexToRgba(theme.palette.black, 0.08)}`,
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: `inset 0 2px ${hexToRgba(theme.palette.black, 0.16)}`,
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: `0px -2px ${hexToRgba(
        theme.palette.black,
        0.16,
      )} inset, 0px 2px ${hexToRgba(theme.palette.black, 0.08)}`,
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: `inset 0 2px ${hexToRgba(theme.palette.black, 0.16)}`,
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.gray.deepDark,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
});

const paddingAtActive: Record<
  ButtonSize,
  { paddingTop: string; paddingBottom: string }
> = {
  small: {
    paddingTop: "7px",
    paddingBottom: "5px",
  },
  medium: {
    paddingTop: "11px",
    paddingBottom: "9px",
  },
  large: {
    paddingTop: "14px",
    paddingBottom: "12px",
  },
};

const verticalPadding: Record<ButtonSize, { padding: string }> = {
  small: {
    padding: "6px",
  },
  medium: {
    padding: "10px",
  },
  large: {
    padding: "13px",
  },
};

const getPadding = ({ theme, size, color }: Padding) => ({
  horizontalPadding: size === "small" ? `10px` : `${theme.spacing * 2}px`,
  verticalPadding: verticalPadding[size].padding,
  paddingTopAtActive: color === "clear" ? "" : paddingAtActive[size].paddingTop,
  paddingBottomAtActive:
    color === "clear" ? "" : paddingAtActive[size].paddingBottom,
});

export type ButtonProps = Omit<BaseButtonProps, "color"> & {
  /**
   * The component used for the root node.
   * Default: `<button />`
   */
  component?:
    | keyof JSX.IntrinsicElements
    | React.ComponentType<{ className: string }>;
  color?: ButtonColor;
  /**
   * Control whether "inline" or "block" Element.
   */
  inline?: boolean;
  size?: ButtonSize;
  onClick?: (event: React.MouseEvent<Element, MouseEvent>) => void;
  /**
   * If added this props, root node becomes `<a />`.
   */
  href?: string;
};

const Button = React.forwardRef<HTMLElement, ButtonProps>(
  (
    {
      component,
      children,
      color = "primary",
      inline = false,
      size = "medium",
      href,
      ...rest
    },
    ref,
  ) => {
    const theme = useTheme();
    const colorStyle = getContainerColorStyles(theme)[color];
    const {
      horizontalPadding,
      verticalPadding,
      paddingTopAtActive,
      paddingBottomAtActive,
    } = getPadding({
      theme,
      size,
      color,
    });

    const isLink = !!href;
    let anchorProps: any = {};
    if (isLink) {
      anchorProps = {
        as: component || "a",
        href,
      };
    }

    return (
      <Styled.ButtonContainer
        ref={ref}
        {...rest}
        as={component || "button"}
        {...anchorProps}
        inline={inline}
        verticalPadding={verticalPadding}
        horizontalPadding={horizontalPadding}
        paddingTopAtActive={paddingTopAtActive}
        paddingBottomAtActive={paddingBottomAtActive}
        normal={{ ...colorStyle.normal }}
        hover={{ ...colorStyle.hover }}
        active={{ ...colorStyle.active }}
        fontWeight={color === "secondary" ? "normal" : "bold"}
        fontSize={
          size === "small" ? `${fontSize["xs"]}px` : `${fontSize["md"]}px`
        }
      >
        {children}
      </Styled.ButtonContainer>
    );
  },
);

export default Button;
