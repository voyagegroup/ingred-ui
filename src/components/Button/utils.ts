import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
