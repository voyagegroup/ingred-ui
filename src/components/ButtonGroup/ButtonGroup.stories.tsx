import * as React from "react";
import { Story } from "@storybook/react/types-6-0";
import ButtonGroup from ".";
import Button from "../Button";
import { ButtonGroupProps } from "./ButtonGroup";
import { Flex, Spacer, Typography } from "..";

export default {
  title: "Components/Inputs/ButtonGroup",
  component: ButtonGroup,
  subcomponents: { Button },
};

export const Example: Story<ButtonGroupProps> = (args) => (
  <ButtonGroup {...args}>
    <Button>Save</Button>
    <Button>Edit</Button>
    <Button>Delete</Button>
    <Button>Cancel</Button>
  </ButtonGroup>
);

export const DesignSamples = () => (
  <Flex display="flex">
    <div>
      <Typography weight="bold">Size Medium</Typography>
      <Spacer pt={2} />
      <ButtonGroup>
        <Button>Save</Button>
        <Button>Edit</Button>
        <Button>Delete</Button>
        <Button>Cancel</Button>
      </ButtonGroup>
    </div>
    <Spacer pl={5} />
    <div>
      <Typography weight="bold">Size Small</Typography>
      <Spacer pt={2} />
      <ButtonGroup size="small">
        <Button>Save</Button>
        <Button>Edit</Button>
        <Button>Delete</Button>
        <Button>Cancel</Button>
      </ButtonGroup>
    </div>
  </Flex>
);

export const DisablePartially: Story<ButtonGroupProps> = (args) => (
  <ButtonGroup {...args}>
    <Button>Save</Button>
    <Button disabled>Edit</Button>
    <Button>Delete</Button>
    <Button>Cancel</Button>
  </ButtonGroup>
);

export const LinkMixed: Story<ButtonGroupProps> = (args) => {
  const Link: React.FC<{ href: string; className: string }> = ({
    href,
    className,
    children,
  }) => (
    // MEMO: Add className props to apply style
    <a href={href} className={className}>
      {children}
    </a>
  );
  return (
    <ButtonGroup {...args}>
      <Button component={Link} href="#">
        Link via {`<Link />`} Component
      </Button>
      <Button href="#">Link via href props</Button>
    </ButtonGroup>
  );
};

LinkMixed.parameters = { docs: { source: { type: "code" } } };
