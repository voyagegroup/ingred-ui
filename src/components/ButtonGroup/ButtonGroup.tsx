import * as React from "react";
import * as Styled from "./styled";

import { ButtonSize } from "../Button/Button";
import { useTheme } from "../../themes";

type GroupButtonSize = Exclude<ButtonSize, "large">;

export type ButtonGroupProps = React.ComponentPropsWithoutRef<"div"> & {
  size?: GroupButtonSize;
  disabled?: boolean;
};

const buttonSize: Record<GroupButtonSize, { minWidth: string }> = {
  small: {
    minWidth: "63px",
  },
  medium: {
    minWidth: "71px",
  },
};

const ButtonGroup = React.forwardRef<HTMLDivElement, ButtonGroupProps>(
  function ButtonGroup(
    { size = "medium", disabled = false, children, ...rest },
    ref,
  ) {
    const theme = useTheme();
    const horizontalPadding =
      size === "small" ? `${theme.spacing}px` : `${theme.spacing * 2}px`;

    const childrenWithProps = React.Children.map(
      children as React.ReactElement[],
      (child: React.ReactElement) => {
        const Button = React.cloneElement(child, {
          ...child.props,
          ...(disabled && { disabled: true }),
          size: size,
          color: "secondary",
          style: {
            ...child.props.style,
          },
        });
        return Button;
      },
    );

    return (
      <Styled.ButtonGroupContainer
        ref={ref}
        minWidth={buttonSize[size].minWidth}
        horizontalPadding={horizontalPadding}
        {...rest}
      >
        {childrenWithProps}
      </Styled.ButtonGroupContainer>
    );
  },
);

export default ButtonGroup;
