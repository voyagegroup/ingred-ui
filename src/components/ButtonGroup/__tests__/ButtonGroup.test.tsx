import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import ButtonGroup from "..";
import Button from "../../Button";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Button component testing", () => {
  afterEach(cleanup);

  test("ButtonGroup", () => {
    const { asFragment } = renderWithThemeProvider(
      <ButtonGroup>
        <Button>Edit</Button>
        <Button>Save</Button>
      </ButtonGroup>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
