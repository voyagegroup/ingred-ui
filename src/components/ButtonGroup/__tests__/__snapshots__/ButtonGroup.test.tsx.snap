// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component testing ButtonGroup 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL hfxIrv"
  >
    <button
      class="sc-dkPtRN sc-hKwDye iYzoKV FiWGy"
      font-size="14px"
      font-weight="normal"
    >
      Edit
    </button>
    <button
      class="sc-dkPtRN sc-hKwDye iYzoKV FiWGy"
      font-size="14px"
      font-weight="normal"
    >
      Save
    </button>
  </div>
</DocumentFragment>
`;
